bpfd-operator/controllers/bpfd-agent/common.go

```diff
@@ -234,109 +235,120 @@ func (r *ReconcilerCommon) createBpfProgram(ctx context.Context,
 // the core logic for taking *Program objects, turning them into bpfProgram
 // object(s), and ultimately telling the custom controller types to load real
 // bpf programs on the node via bpfd. Additionally it acts as a central point for
-// interacting with the K8s API.
+// interacting with the K8s API. This function will exit if any action is taken
+// against the K8s API.
 func reconcileProgram(ctx context.Context,
 	rec bpfdReconciler,
 	program client.Object,
 	common *bpfdiov1alpha1.BpfProgramCommon,
 	ourNode *v1.Node,
 	programMap map[string]*gobpfd.ListResponse_ListResult) (bool, error) {
 
+	// initialize reconciler state
 	r := rec.getRecCommon()
-	r.bpfPrograms = map[string]bpfdiov1alpha1.BpfProgram{}
 
-	// Populate already existing bpfPrograms for *Program.
-	err := r.getBpfPrograms(ctx, program)
+	// determine which node local actions should be taken based on wether the node is selected
+	// OR if the *Program is being deleted.
+	isNodeSelected, err := isNodeSelected(&common.NodeSelector, ourNode.Labels)
 	if err != nil {
-		return false, fmt.Errorf("failed to get bpfPrograms: %v", err)
+		return false, fmt.Errorf("failed to check if node is selected: %v", err)
 	}
 
-	// Get and Create any new bpfPrograms exiting after each creation since
-	// the K8s API "create" action will re-trigger the reconcile loop allowing the
-	// logic to continue incrementally.
-	progList, err := rec.buildBpfPrograms(ctx)
-	if err != nil {
-		r.Logger.Error(err, "failed to create bpfPrograms")
-		return true, nil
-	}
+	isBeingDeleted := !program.GetDeletionTimestamp().IsZero()
 
-	for _, prog := range progList.Items {
-		_, exists := r.bpfPrograms[prog.Name]
-		if !exists {
-			opts := client.CreateOptions{}
-			r.Logger.Info("creating bpfProgram", "Name", prog.Name)
-			if err := r.Create(ctx, &prog, &opts); err != nil {
-				return true, fmt.Errorf("failed to create bpfProgram object: %v", err)
-			}
-			r.bpfPrograms[prog.Name] = prog
-			return false, nil
-		}
+	// Get existing bpfPrograms for a *Program.
+	existingPrograms, err := r.getExistingBpfProgs(ctx, program)
+	if err != nil {
+		return false, fmt.Errorf("failed to get bpfPrograms: %v", err)
 	}
 
-	isNodeSelected, err := isNodeSelected(&common.NodeSelector, ourNode.Labels)
+	// Get expected bpfPrograms for a *Program.
+	expectedPrograms, err := rec.expectedBpfPrograms(ctx)
 	if err != nil {
-		return false, fmt.Errorf("failed to check if node is selected: %v", err)
+		r.Logger.Error(err, "failed to create bpfPrograms")
+		return true, nil
 	}
 
-	isBeingDeleted := !program.GetDeletionTimestamp().IsZero()
-
+	// Get program bytecode, source could be an OCI container image or filepath
 	bytecode, err := bpfdagentinternal.GetBytecode(r.Client, &common.ByteCode)
```

Comment:
The reset of the functions above "should just work", but this one can fail if the authentication is incorrect.  As an optimization, this could be moved to the top of the function so it can fail early.